### PR TITLE
TL/UCP: Add all-reduce ring alogorithm

### PR DESCRIPTION
**Pull request by armratner**
## What
This PR adds a new ring-based Allreduce algorithm (named `"ring"`) to the UCP transport layer within UCC. It introduces:
- A new source file `allreduce_ring.c` implementing the ring-based method.
- Modifications to the build system (`Makefile.am`) to include the new file.
- Updates to the Allreduce interface (`allreduce.[ch]`), including a new enum value `UCC_TL_UCP_ALLREDUCE_ALG_RING`, new function prototypes, and references in the algorithm registration.
- The ring-based algorithm’s logic (init, start, progress, and finalize) in `allreduce_ring.c` that manages per-rank scratch buffers, chunk-based sending/receiving, and reduction.

## Why ?
A ring-based Allreduce can be more efficient for large message sizes, especially on relatively simple or homogeneous network topologies. It complements existing Allreduce algorithms (e.g., knomial, sliding window, DBT) by providing:
- Improved scalability for certain message sizes.
- A straightforward method for ring-style communication patterns common in distributed HPC and AI workloads.

## How ?
The ring algorithm splits the input data into chunks, then circulates these chunks around the ring of ranks. Each rank performs local partial reductions on received data and passes it along. The main changes include:

1. **File Additions/Modifications**:  
   - `allreduce_ring.c`: Implements the ring-based send/recv steps, in-place or out-of-place usage, and partial data reductions via `ucc_dt_reduce`.  
   - `Makefile.am`: Includes the new file in the build.  
   - `allreduce.c/allreduce.h`: Adds the new `"ring"` algorithm ID and associated function prototypes.

2. **Implementation Details**:  
   - Data is divided into `num_chunks`, typically equal to the number of ranks. Each chunk is passed around the ring (`sendto`/`recvfrom`) and reduced in a scratch buffer.  
   - A `scratch` buffer is allocated per rank to hold incoming chunk data before reduction.  
   - The algorithm ensures all chunks complete one round in the ring, then finalizes once the entire data is fully reduced on each rank.

3. **Code Flow**:  
   - **Init**: Sets up the ring task, scratch buffer, and references to the team’s executor.  
   - **Start**: Posts initial sends/receives and enqueues the progress function.  
   - **Progress**: Drives the ring of sends/receives chunk by chunk, calling `ucc_dt_reduce` on each incoming portion.  
   - **Finalize**: Cleans up (frees scratch space and finishes the task).

By adding this ring-based approach, UCC gains a more complete suite of collective algorithms for Allreduce, allowing users and internal heuristics to pick the best method based on message size, topology, and system capabilities.

### CODE REVIEW

**Comment by armratner** [comment]
Working on Gtest

File: src/components/tl/ucp/allreduce/allreduce.h [hunk 1]
```diff
@@ -77,6 +78,18 @@ ucc_status_t ucc_tl_ucp_allreduce_dbt_start(ucc_coll_task_t *task);
 
 ucc_status_t ucc_tl_ucp_allreduce_dbt_progress(ucc_coll_task_t *task);
 
+void ucc_tl_ucp_allreduce_ring_progress(ucc_coll_task_t *coll_task);
+
+ucc_status_t ucc_tl_ucp_allreduce_ring_start(ucc_coll_task_t *coll_task);
+
+ucc_status_t ucc_tl_ucp_allreduce_ring_init_common(ucc_tl_ucp_task_t *task);
```
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/allreduce/allreduce.h:85]: Removed ucc_tl_ucp_allreduce_ring_init_common from the header file since it's only used internally.

File: src/components/tl/ucp/tl_ucp_coll.h [hunk 2]
```diff
@@ -269,6 +269,12 @@ typedef struct ucc_tl_ucp_task {
             int                     phase;
         } alltoall_bruck;
         char                        plugin_data[UCC_TL_UCP_TASK_PLUGIN_MAX_DATA];
+        struct {
+            void                   *scratch;
+            ucc_mc_buffer_header_t *scratch_mc_header;
+            ucc_ee_executor_task_t *etask;
+            ucc_ee_executor_t      *executor;
+        } allreduce_ring;
```
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/tl_ucp_coll.h:277]: Rearranged the order of structs in tl_ucp_coll.h to keep plugin_data at the end.

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                                   task->allreduce_ring.executor,
+                                   &task->allreduce_ring.etask);
+            if (ucc_unlikely(status != UCC_OK)) {
+                tl_error(UCC_TASK_LIB(task), "failed to perform dt reduction");
+                task->super.status = status;
+                return;
+            }
+        }
+
+        task->tagged.send_posted++;
```
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/allreduce/allreduce_ring.c:76]: Yes this is defiantly a bug, it should increment a dedicated step counter which helps track the ring operations.
I fixed it.

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    while (task->tagged.send_posted < tsize - 1) {
+        int step = task->tagged.send_posted;
+
+        for (int chunk = 0; chunk < num_chunks; chunk++) {
+            offset = chunk * chunk_size;
+            remaining = (chunk == num_chunks - 1) ? data_size - offset : chunk_size;
+
+            send_buf  = (step == 0) ? sbuf + offset : rbuf + offset;
+            recv_buf  = task->allreduce_ring.scratch + offset;
```
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/allreduce/allreduce_ring.c:50]: Fixed

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    size_t             count     = TASK_ARGS(task).dst.info.count;
+    ucc_datatype_t     dt        = TASK_ARGS(task).dst.info.datatype;
+    size_t             data_size = count * ucc_dt_size(dt);
+    size_t             chunk_size, offset, remaining;
+    ucc_rank_t         sendto, recvfrom;
+    void              *recv_buf, *send_buf, *reduce_buf;
+    ucc_status_t       status;
+
+    int num_chunks = tsize; // Use the number of ranks as the number of chunks (this is dynamic)
+    chunk_size = (data_size + num_chunks - 1) / num_chunks; // Ensure chunks fit into data evenly
```
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/allreduce/allreduce_ring.c:29]: Enhanced comment

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    size_t             count     = TASK_ARGS(task).dst.info.count;
+    ucc_datatype_t     dt        = TASK_ARGS(task).dst.info.datatype;
+    size_t             data_size = count * ucc_dt_size(dt);
+    ucc_status_t       status;
+
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_allreduce_ring_start", 0);
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    /* Allocate scratch space for the receive buffer */
+    status = ucc_mc_alloc(&task->allreduce_ring.scratch_mc_header,
```
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/allreduce/allreduce_ring.c:98]: fixed

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    if (!(task->flags & UCC_TL_UCP_TASK_FLAG_SUBSET)) {
+        if (team->cfg.use_reordering) {
+            sbgp = ucc_topo_get_sbgp(team->topo, UCC_SBGP_FULL_HOST_ORDERED);
+            task->subset.myrank = sbgp->group_rank;
+            task->subset.map    = sbgp->map;
+        }
+    }
+
+    task->super.post     = ucc_tl_ucp_allreduce_ring_start;
+    task->super.progress = ucc_tl_ucp_allreduce_ring_progress;
```
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/allreduce/allreduce_ring.c:128]: Fixed

File: src/components/tl/ucp/tl_ucp_coll.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (armratner) [reply src/components/tl/ucp/tl_ucp_coll.h:277]: Fixed

**Comment by armratner** [comment]
Added the gtest, 

1. Data Type and Operation Coverage
      - INT32 with sum operation
      - FLOAT32 with sum operation
      - INT32 with product operation
      - INT32 with max operation
      - INT32 with min operation
      - FLOAT64 with sum operation
  2. Standard Test (ring)
    - Tests the ring algorithm with standard configurations:
    - Different data sizes (8, 65536, 123567 elements)
    - Both in-place and non-in-place operations
    - Different memory types (HOST, CUDA, CUDA_MANAGED where available)
    - 3 iterations per configuration to ensure stability
  3. Edge Case Test (ring_edge_cases)
    - Tests the ring algorithm with non-power-of-two team sizes (3, 7, 13)
    - Tests edge cases with message sizes:
    - Empty message (0 elements)
    - Single element (1 element)
    - Small odd sizes (3, 17 elements)
  4. Persistent Operation Test (ring_persistent)
    - Tests the algorithm's behavior in persistent mode
    - Uses a consistent buffer size (1024 elements)
    - Runs 5 iterations to verify consistency

File: src/components/tl/ucp/allreduce/allreduce.c [hunk 8]
```diff
@@ -29,6 +29,10 @@ ucc_base_coll_alg_info_t
             {.id   = UCC_TL_UCP_ALLREDUCE_ALG_SLIDING_WINDOW,
              .name = "sliding_window",
              .desc = "sliding window allreduce (optimized for running on DPU)"},
+        [UCC_TL_UCP_ALLREDUCE_ALG_RING] =
+            {.id   = UCC_TL_UCP_ALLREDUCE_ALG_RING,
+             .name = "ring",
+             .desc = "ring-based allreduce (optimized for BW and simple topologies)"},
```
Comment (samnordmann) [inline src/components/tl/ucp/allreduce/allreduce.c:35]: At least for me, "simple topology" doesn't refer to something clear.
I would personally only keep "optimized for BW", or would use "ring topology"

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            
+            if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+                /* Return and resume from this chunk next time */
+                return;
+            }
+
+            status = ucc_dt_reduce(send_buf, recv_buf, reduce_buf,
+                                   remaining / ucc_dt_size(dt),
+                                   dt, &TASK_ARGS(task), 0, 0,
+                                   task->allreduce_ring.executor,
```
Comment (samnordmann) [inline src/components/tl/ucp/allreduce/allreduce_ring.c:79]: the executor has not been initialiazed

**Review by samnordmann — COMMENTED** [review]
LGTM, however I do not see the tests passing the CI logs. Can you push a new commit to Trigger the CI? @janjust do we know why the Jenkin CI hasnt been triggered?

I think the algo might run into a segfault because the executor is not initialized

**Review by nsarka — COMMENTED** [review]
Thanks for the alg!

**Comment by janjust** [comment]
@janjust do we know why the Jenkin CI hasnt been triggered?
Not sure - we had issues with jenkins for a bit now, but working with Artem and Andrii to resolve it - should be good now, we can just retrigger.

File: src/components/tl/ucp/allreduce/allreduce.h [hunk 10]
```diff
@@ -77,6 +78,16 @@ ucc_status_t ucc_tl_ucp_allreduce_dbt_start(ucc_coll_task_t *task);
 
 ucc_status_t ucc_tl_ucp_allreduce_dbt_progress(ucc_coll_task_t *task);
 
+void ucc_tl_ucp_allreduce_ring_progress(ucc_coll_task_t *coll_task);
+
+ucc_status_t ucc_tl_ucp_allreduce_ring_start(ucc_coll_task_t *coll_task);
```
Comment (ikryukov) [inline src/components/tl/ucp/allreduce/allreduce.h:83]: Advice: Consider not exposing functions in the public API if they are not intended for use outside the algorithm's internal logic. Keeping the API surface minimal helps with maintainability and reduces potential misuse.

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    /* 
+     * In the ring algorithm, each process sends/receives tsize-1 times
+     * This is because after tsize-1 steps, each piece of data has traversed 
+     * the entire ring and completed its reduction
+     */
+    while (task->allreduce_ring.step < tsize - 1) {
+        step = task->allreduce_ring.step;
+        
+        /* Resume from the last processed chunk */
+        for (chunk = task->allreduce_ring.chunk; chunk < num_chunks; chunk++) {
```
[the remark this answers is not in the document]
Comment (nsarka) [reply src/components/tl/ucp/allreduce/allreduce_ring.c:53]: I think doing it with 2 loops would also reduce the scratch buffer size to just `chunk_size` as well, instead of the size of the recv buf

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+    void              *sbuf      = TASK_ARGS(task).src.info.buffer;
+    void              *rbuf      = TASK_ARGS(task).dst.info.buffer;
+    ucc_memory_type_t  mem_type  = TASK_ARGS(task).dst.info.mem_type;
+    size_t             count     = TASK_ARGS(task).dst.info.count;
+    ucc_datatype_t     dt        = TASK_ARGS(task).dst.info.datatype;
+    size_t             dt_size   = ucc_dt_size(dt);
+    size_t             data_size = count * dt_size;
+
+    /* Early return for zero-count or single-rank edge cases */
+    if (data_size == 0 || tsize <= 1) {
```
Comment (nsarka) [inline src/components/tl/ucp/allreduce/allreduce_ring.c:27]: Should this be in start?

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+        return;
+    }
+
+    size_t             chunk_size, send_offset, recv_offset;
+    size_t             send_chunk_size, recv_chunk_size;
+    ucc_rank_t         sendto, recvfrom;
+    void              *recv_buf, *send_buf, *reduce_buf;
+    ucc_status_t       status;
+    int                step;
+    int                num_chunks = tsize;
```
Comment (nsarka) [inline src/components/tl/ucp/allreduce/allreduce_ring.c:43]: Initialized variables should be declared first

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+    /* Single-phase Ring Algorithm (SRA):
+     * - Each rank starts with its local chunk fully reduced
+     * - In each step, ranks exchange chunks and combine them
+     * - After tsize-1 steps, all ranks have the complete reduced result
+     */
+
+    /* On first entry, initialize step and phase */
+    if (task->allreduce_ring.step == 0 && task->allreduce_ring.phase == RING_PHASE_INIT) {
+        if (!UCC_IS_INPLACE(TASK_ARGS(task))) {
+            memcpy(rbuf, sbuf, data_size);
```
Comment (nsarka) [inline src/components/tl/ucp/allreduce/allreduce_ring.c:72]: Should this happen in start?

File: test/gtest/coll/test_allreduce.cc [hunk 15]
```diff
@@ -437,6 +445,95 @@ TYPED_TEST(test_allreduce_alg, rab_pipelined) {
...
+            this->data_fini(ctxs);
+        }
+    }
+}
+
+TYPED_TEST(test_allreduce_alg, ring_persistent) {
+    // Test persistent operation - results should be consistent across multiple calls
+    int           n_procs = 8;
+    ucc_job_env_t env     = {{"UCC_CL_BASIC_TUNE", "inf"},
+                             {"UCC_TL_UCP_ALLREDUCE_ALG", "ring"}};
```
Comment (nsarka) [inline test/gtest/coll/test_allreduce.cc:515]: It should be `UCC_TL_UCP_TUNE=allreduce:0-inf:@ring`

**Review by nsarka — COMMENTED** [review]
Please double check the gtest, it may not have been running during the test based on the env

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+            }
+
+            /* Calculate receive offset and chunk size */
+            recv_offset = recv_chunk * chunk_size;
+            if (recv_offset >= data_size) {
+                task->allreduce_ring.phase = RING_PHASE_COMPLETE;
+                continue;
+            }
+
+            recv_chunk_size = data_size - recv_offset;
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/allreduce/allreduce_ring.c:146]: use ucc_block_size

File: src/components/tl/ucp/allreduce/allreduce_ring.c [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+    *task_h = &task->super;
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_allreduce_ring_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_status_t st, global_st = UCC_OK;
+
+    if (task->allreduce_ring.etask != NULL) {
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/allreduce/allreduce_ring.c:301]: etask should always be null in finalize, otherwise collective is not finished

**Comment by armratner** [comment]
| Message Size
Emphasising the Latency Jump at 8 MB
 (MB) | UCC Ring (µs) | UCC Knomial (µs) | OMPI Tuned (µs) |
|------------------:|--------------:|-----------------:|----------------:|
| 1                                |     2 287.61 |         579.78 |      3 710.72 |
| 2                               |     2 656.15 |         983.58 |      7 397.25 |
| 4                               |     4 380.01 |       3 531.90 |     15 340.96 |
| 8                               |    11 228.21 |      12 728.49 |     32 010.70 |
| 16                             |    26 984.31 |      31 163.98 |     70 183.02 |
| 32                             |    50 100.87 |      69 753.03 |    154 048.94 |
| 64                             |   100 628.53 |     145 987.70 |    314 291.89 |
| 128                           |   200 764.59 |     295 283.81 |    638 286.61 |
| 256                          |   405 462.40 |     590 662.48 |         —      |
| 512                           |   992 232.12 |   1 182 699.49 |         —      |

**Comment by armratner** [comment]
Scale context – 128 ranks / 4 nodes / 32 PPN

**Comment by janjust** [comment]
Closing this in favor of #1258
